lib/type.ml

```diff
@@ -22,15 +26,124 @@ type constraint_ =
+  | TySingleton (Number n) :: ty1s when List.exists ~f:((=) TyNumber) store ->
+     sup_elems_to_list store ty1s
+  | TySingleton (Number n) :: ty1s ->
+     sup_elems_to_list (TySingleton (Number n) :: store) ty1s
+  | TyAtom :: ty1s ->
+     let is_not_atom = function TySingleton (Atom _) -> false | _ -> true in
+     let store' = TyAtom :: List.filter ~f:is_not_atom store in
+     sup_elems_to_list store' ty1s
+  | TySingleton (Atom a) :: ty1s when List.exists ~f:((=) TyAtom) store ->
+     sup_elems_to_list store ty1s
+  | TySingleton (Atom a) :: ty1s ->
+     sup_elems_to_list (TySingleton (Atom a) :: store) ty1s
+  | TyTuple ty2s :: ty1s ->
+     let store' =
+       if List.exists ~f:(function TyTuple tys when List.length ty2s = List.length tys -> true | _ -> false) store then
+         List.map ~f:(function
+                      | TyTuple tys when List.length ty2s = List.length tys ->
+                         List.map2_exn ~f:sup tys ty2s
+                         |> fun ty2s' -> TyTuple ty2s'
+                      | t -> t)
+                  store
+       else
+         TyTuple ty2s :: store
+     in
+     sup_elems_to_list store' ty1s
+  | TyFun (args, range) :: ty1s ->
+     let store' =
+       if List.exists ~f:(function TyFun (args0, _) when List.length args0 = List.length args -> true | _ -> false) store then
+         List.map ~f:(function
+                      | TyFun (args0, range0) when List.length args0 = List.length args ->
+                         List.map2_exn ~f:sup args0 args
+                         |> fun ty2s' -> TyTuple ty2s'
+                      | t -> t)
+                  store
+       else
+         TyFun (args, range) :: store
+     in
+     sup_elems_to_list store' ty1s
+
+let union_list tys =
+  List.reduce_exn ~f:sup tys
+
+let rec inf ty1 ty2 =  (* ty1 and ty2 should be a TyAny, TyBottom, TyVar or TyUnion *)
+  match (ty1, ty2) with
+  | _ when ty1 = ty2 -> ty1
+  | (TyAny, ty) | (ty, TyAny) -> ty
+  | (TyBottom, _) | (_, TyBottom) -> TyBottom
+  | (TyUnion tys1, TyUnion tys2) ->
+     let ty1s =
+       List.cartesian_product tys1 tys2
+       |> List.filter_map ~f:(fun (ty1, ty2) -> inf_elem ty1 ty2)
+     in
+     begin match sup_elems_to_list [] ty1s with
+     | [] -> TyBottom
+     | ty1s -> TyUnion ty1s
+     end
+and inf_elem ty1 ty2 =  (* ty1 and ty2 should be a TyNumber, TySingleton, TyAtom, TyTuple or TyFun *)
+  match (ty1, ty2) with
+  | (TyVar _, _) | (_, TyVar _) -> failwith "cannot reach here"
+  | _ when ty1 = ty2 -> Some ty1
+  | (TyNumber, TySingleton (Number n)) | (TySingleton (Number n), TyNumber) ->
+     Some (TySingleton (Number n))
+  | (TyAtom, TySingleton (Atom a)) | (TySingleton (Atom a), TyAtom) ->
+     Some (TySingleton (Atom a))
+  | (TyTuple tys1, TyTuple tys2) when List.length tys1 = List.length tys2 ->
+     List.map2_exn ~f:inf tys1 tys2
+     |> fun tys -> Some (TyTuple tys)
+  | (TyFun (args1, range1), TyFun (args2, range2)) ->
```

Comment:
Is `when List.length args1 = List.length args2` missing?

Reply:
Yes, thank you. fixed it